File: haskell/pig-latin/PigLatin.hs
module PigLatin (translate) where

import Control.Applicative ((<|>), (<$>))
import Data.Maybe (fromMaybe)
import Data.List (isPrefixOf, find)

translate :: String -> String
translate = unwords . map translateWord . words

translateWord :: String -> String
translateWord word = fromMaybe (translateSingleLetterConsonantWord word) 
	(translateVowelWord word <|> 
	 translateMultiLetterConsonantWord word <|> 
	 translateQuConsonantWord word)

translateVowelWord :: String -> Maybe String
translateVowelWord word = mapPrefix (vowelWordToPigLatin word) ["a", "e", "i", "u", "yt", "xr"] word 
	
vowelWordToPigLatin :: String -> String -> String
vowelWordToPigLatin word vowel = vowel ++ drop (length vowel) word ++ "ay"

translateMultiLetterConsonantWord :: String -> Maybe String
translateMultiLetterConsonantWord word = mapPrefix (consonantWordToPigLatin word) ["ch", "qu", "thr", "th", "sch", "yt"] word

translateSingleLetterConsonantWord :: String -> String
translateSingleLetterConsonantWord word = drop 1 word ++ take 1 word ++ "ay"

translateQuConsonantWord :: String -> Maybe String
translateQuConsonantWord word 
	| "qu" == take 2 (drop 1 word) = Just $ drop 3 word ++ take 1 word ++ "quay"
	| otherwise = Nothing

consonantWordToPigLatin :: String -> String -> String
consonantWordToPigLatin word consonant = drop (length consonant) word ++ consonant ++ "ay"

mapPrefix :: (String -> String) -> [String] -> String -> Maybe String
mapPrefix convert prefixes word = convert <$> find (`isPrefixOf` word) prefixes

File: clojure/pig-latin/src/pig_latin.clj
(ns pig-latin (:require [clojure.string :as str]))

(def vowel-regex #"(?<begin>^|\s+)(?<vowel>a|e|i|o|u|yt|xr)(?<rest>\w+)")
(def consonant-regex #"(?<begin>^|\s+)(?<consonant>ch|qu|thr|th|sch|yt|\wqu|\w)(?<rest>\w+)")

(def vowel-replacement "${begin}${vowel}${rest}ay")
(def consonant-replacement "${begin}${rest}${consonant}ay")

(defn translate [word]
  (if (re-matches vowel-regex word)
    (str/replace word vowel-regex vowel-replacement)
    (str/replace word consonant-regex consonant-replacement)))
